crossplane/crossplane
# Update default registry to xpkg.upbound.io

### Description of your changes

This PR updates the default value for the registry that Crossplane's package manager will install packages from. The default was previously `index.docker.io`, and the new default will be `xpkg.upbound.io`.  Details about the motivation for this change can be ready in #5012.

In addition to the simple change of the default value, it was necessary to update the package manager logic that creates deployment manifests that run providers and functions. That logic is now aware of the default registry so it can include it into manifests it creates when the user has not specified an explicit registry source. 

Feedback is welcome ASAP so we can gain confidence that these changes are thorough and we aren't missing any scenarios.

Fixes #5012 

### Testing

The following scenarios have been tested so far and their behavior is described briefly.  Links to gists with specific testing steps are provided for the more complicated scenarios in the table.

| Pkg Registry Set | Default Set | Upgrade Performed | Scenario | Behavior | 
| --------------------------- | ------------------| -------------------------------- | --------- | -------- |
| ❌  | ❌  | ❌  | Install new package w/o explicit registry specified: Providers, Functions, and Configurations | ✅ Package is installed from xpkg.upbound.io, can create resources OK |
| ✅  | ❌  | ❌  | Install new package with explicit registry specified  | ✅ Package is installed from specified registry, can create resources OK | 
| ❌  | ❌  | ✅  | Package installed on v1.14 w/o explicit registry specified, then Crossplane upgraded to this PR</br>Steps: [gist link](https://gist.github.com/jbw976/793210f26f2dfbb458a41d5dc437ce7e) | ⚠️  Behavior change: the package manager will attempt to update the existing package to the new default registry location. If the package does not exist on the new registry location, the package will become unhealthy. | 
| ✅  | ❌  | ✅  | Package installed on v1.14 w/ explicit registry specified, then Crossplane upgraded to this PR</br>Steps: [gist link](https://gist.github.com/jbw976/83695c4a31f7149bfbaea8249ed638c8) | ✅ Package stays healthy after the upgrade, existing resources stay healthy, can create new resources OK | 
| ❌  | ✅  | ✅  | Package installed on v1.14 w/o explicit registry specified, then Crossplane is upgraded to this PR with `--registry=index.docker.io` to keep previous behavior</br>[gist link](https://gist.github.com/jbw976/094cdfab0eb502e8369259f605761bda) | ✅ The behavior from previous releases of Crossplane remains intact, package stays healthy and installed from dockerhub. |
| ❌  | ❌  | ❌  | Dependencies in a package that are specified w/o an explicit registry [gist link](https://gist.github.com/jbw976/afd49f0a5fdc59a2c7aff559b00fb5ac) | ✅   Package and its dependency are installed from the default registry OK and everything reaches a healthy state. | 
| ✅  | ❌  | ❌  | `DeploymentRuntimeConfig` specifies an image w/ no registry set that overrides the image the package was installed with [gist link](https://gist.github.com/jbw976/f518c7c8f2d1d384c88f75f078d5119a)  | ✅  The image specified in the `DeploymentRuntimeConfig` is used exactly as is, no default is taken into account. If this image doesn't exist then the package will not get healthy, but we will honor exactly what the user specifies in the DRC. | 
| ❌  | ❌  | ❌  | `ControllerConfig` specifies an image w no registry set that overrides the image the package was installed with [gist link](https://gist.github.com/jbw976/fff6dc20c1042079182e55579fa5a1a2) | ✅  The image specified in the `ControllerConfig` is used exactly as is, no default is taken into account. Same behavior as `DeploymentRuntimeConfig`. | 


The most challenging behavior will be the likely common scenario of having existing packages installed implicitly from dockerhub (by previously not explicitly specifying a registry), then upgrading Crossplane to the functionality in this PR. Since the package manager will essentially try to upgrade the package from its previous location (dockerhub) to its new location on the default registry (xpkg.upbound.io), the package can become unhealthy if it doesn't exist there.

**Note** that this will happen for all the `crossplanecontrib` packages on dockerhub because that organization is called `crossplane-contrib` on `xpkg.upbound.io` (note the lack of a `-` on dockerhub), and therefore those packages won't exist.

We will absolutely need clear messaging about this change in the release notes and we should write a blog post too.  For anyone that has an environment where this behavior will cause problems, **ANY** of these options would be recommended:

* update your packages so they all have an explicit registry specified, e.g. `index.docker.io/myorg/mypkg`
* copy/push your packages to `xpkg.upbound.io` so they exist there in addition to dockerhub
* keep the previous behavior by using `--set args='{"--registry=index.docker.io"}'` during the helm upgrade of Crossplane v1.15

I have: 

- [x] Linked a PR or a [docs tracking issue] to [document this change]: https://github.com/crossplane/docs/issues/680

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet
